### PR TITLE
Uninstall should be amazing

### PR DESCRIPTION
**Pull request by shiftkey**
Resolving #74 

Fixes:
- a stupid enumeration thing I introduced
- moved some diagnostics out to debug to slim down normal logs

TODO:
- [x] on uninstall, deleting the directory supposedly fails as it's not empty, but upon inspection the folder is empty (race condition?)

```
ERROR | 2013-09-10 01:27:46 PM +09 | Utility: DeleteDirectory: could not delete -
 C:\Users\brendanforster\AppData\Local\TestApp: System.IO.IOException: The directory is not empty.

   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, 
          Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, 
          Boolean checkHost)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at Shimmer.Core.Utility.DeleteDirectory(String directoryPath)
WARN | 2013-09-10 01:27:46 PM +09 | UpdateManager: Full Uninstall tried to delete root
 dir but failed, punting until next reboot: System.IO.IOException: The directory is not empty.

   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, 
        Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, 
        Boolean checkHost)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at Shimmer.Core.Utility.DeleteDirectory(String directoryPath)
   at Shimmer.Client.UpdateManager.<fullUninstall>b__26()
ERROR | 2013-09-10 01:27:46 PM +09 | Utility: safeDeleteFileAtNextReboot: failed -
  C:\Users\brendanforster\AppData\Local\TestApp
INFO | 2013-09-10 01:27:49 PM +09 | InstallManager: Full uninstall OnNext: ()
INFO | 2013-09-10 01:27:49 PM +09 | InstallManager: Full uninstall OnCompleted
```

### CODE REVIEW

**Comment by shiftkey** [comment]
Ah that's better: https://gist.github.com/shiftkey/c399f2cc2ee42581364f

File: src/Shimmer.Client/UpdateManager.cs [hunk 1]
```diff
@@ -396,7 +396,7 @@ List<string> installPackageToAppDir(UpdateInfo updateInfo, ReleaseEntry release)
 
             // NB: This might happen if we got killed partially through applying the release
             if (target.Exists) {
-                Utility.DeleteDirectory(target.FullName);
+                Utility.DeleteDirectory(target.FullName).Wait();
```
Comment (shiftkey) [inline src/Shimmer.Client/UpdateManager.cs:398]: Feels icky, but I'd need to make even more changes to this method (which I haven't seen have problems with so far).

File: src/Shimmer.Client/UpdateManager.cs [hunk 2]
```diff
@@ -433,7 +433,7 @@ List<string> installPackageToAppDir(UpdateInfo updateInfo, ReleaseEntry release)
 
         List<string> runPostInstallAndCleanup(Version newCurrentVersion, bool isBootstrapping)
         {
-            log.Debug(CultureInfo.InvariantCulture, "AppDomain ID: {0}", AppDomain.CurrentDomain.Id);
```
Comment (shiftkey) [inline src/Shimmer.Client/UpdateManager.cs:442]: Dead code

File: src/Shimmer.Client/UpdateManager.cs [hunk 3]
```diff
@@ -445,9 +445,6 @@ List<string> runPostInstallAndCleanup(Version newCurrentVersion, bool isBootstra
 
         List<string> runPostInstallOnDirectory(string newAppDirectoryRoot, bool isFirstInstall, Version newCurrentVersion, IEnumerable<ShortcutCreationRequest> shortcutRequestsToIgnore)
         {
-            shortcutRequestsToIgnore.ForEach(x =>
```
Comment (shiftkey) [inline src/Shimmer.Client/UpdateManager.cs:454]: Kids, be careful about enumerating your collections unnecessarily...

File: src/Shimmer.Core/Utility.cs [hunk 4]
```diff
@@ -165,24 +166,44 @@ public static void DeleteDirectory(string directoryPath)
...
+                dir => {
+                    Log().Debug("Recurse into subdirectory: {0}", dir);
+                    return DeleteDirectory(Path.Combine(directoryPath, dir))
+                                .Retry(3);
+                }
+            );
+
+            var subject = new AsyncSubject<Unit>();
+
+            fileOperations
```
Comment (shiftkey) [inline src/Shimmer.Core/Utility.cs]: Throwing myself at the mercy of the world if you know a better way to do this with Rx:
- if deleting a file fails, it should propagate to subscribers
- if all the files and subdirectories are deleted successfully, delete this current directory
- then propagate success to subscribers

EDIT: s/success/completed

Comment (shiftkey) [inline src/Shimmer.Core/Utility.cs]: if you guys are up for a problem solving challenge @paulcbetts @haacked @niik @half-ogre I'd love to hear your thoughts on that code.

**Comment by peters** [comment]
This is probably fixed already, but just in case it's not.. i'd like to mention that during uninstall the application shortcuts created during install are not removed. (0.6.21)

**Comment by shiftkey** [comment]
@peters those are driven by AppSetups specifically, not this behaviour (which is cleaning up the application directory).

Once I start documenting and writing samples for AppSetup I'll ensure that these shortcuts are correctly tidied up.

File: src/Shimmer.Tests/Core/UtilityTests.cs [hunk 5]
```diff
@@ -22,7 +24,72 @@ public void ShaCheckShouldBeCaseInsensitive()
...
+                sw.Start();
+                Utility.DeleteDirectory(tempDir).Wait();
+                sw.Stop();
+                this.Log().Info("Delete took {0}ms", sw.ElapsedMilliseconds);
+
+                Assert.False(Directory.Exists(tempDir));
+            }
+        }
+
+        static void CreateSampleDirectory(string directory)
```
Comment (shiftkey) [inline src/Shimmer.Tests/Core/UtilityTests.cs:56]: This is just a way to create an arbitrary folder structure for shits and giggles

File: src/Shimmer.Client/UpdateManager.cs [hunk 6]
```diff
@@ -214,18 +214,18 @@ public IObservable<Unit> FullUninstall()
...
-                    return;
-                } catch (Exception ex) {
-                    log.WarnException("Full Uninstall tried to delete root dir but failed, punting until next reboot", ex);
-                }
-                
-                Utility.DeleteDirectoryAtNextReboot(rootAppDirectory);
-            }, RxApp.TaskpoolScheduler);
+            var maxVersion = new Version(255, 255, 255, 255);
+            return
+                Observable.Start(() => cleanUpOldVersions(maxVersion))
```
Comment (anaisbetts) [inline src/Shimmer.Client/UpdateManager.cs]: Always pass a scheduler into `Observable.Start` or any other method that has an `IScheduler` Parameter - in this case we want `RxApp.TaskpoolScheduler`

File: src/Shimmer.Client/UpdateManager.cs [hunk 7]
```diff
@@ -214,18 +214,18 @@ public IObservable<Unit> FullUninstall()
...
+                .SelectMany(_ => Utility.DeleteDirectory(rootAppDirectory))
+                .ObserveOn(RxApp.TaskpoolScheduler)
```
Comment (anaisbetts) [inline src/Shimmer.Client/UpdateManager.cs]: Nuke this, since you're already on the task pool

File: src/Shimmer.Client/UpdateManager.cs (hunk 1 above)
Comment (anaisbetts) [inline src/Shimmer.Client/UpdateManager.cs:398]: Are you sure that you're on a background thread in this code?

File: src/Shimmer.Core/Utility.cs (hunk 4 above)
Comment (anaisbetts) [inline src/Shimmer.Core/Utility.cs]: Check out the `MapReduce` method here:

``` cs
[1,2,3].ToObservable()
    .MapReduce(x => { /* Do some Shit */ });
```

File: src/Shimmer.Client/UpdateManager.cs (hunk 1 above)
Comment (shiftkey) [inline src/Shimmer.Client/UpdateManager.cs:398]: Only usage of `installPackageToAppDir`: 

```
ret = await Observable.Start(() => installPackageToAppDir(updateInfo, release), RxApp.TaskpoolScheduler);
progress.OnNext(95);
```

Comment (anaisbetts) [inline src/Shimmer.Client/UpdateManager.cs:398]: That'll do it

File: src/Shimmer.Client/UpdateManager.cs [hunk 8]
```diff
@@ -618,8 +614,9 @@ IObservable<Unit> cleanDeadVersions(Version currentVersion)
             return di.GetDirectories().ToObservable()
                 .Where(x => x.Name.ToLowerInvariant().Contains("app-"))
                 .Where(x => currentVersion != null ? x.Name != getDirectoryForRelease(currentVersion).Name : true)
-                .MapReduce(x => Observable.Start(() => Utility.DeleteDirectory(x.FullName), RxApp.TaskpoolScheduler)
-                    .LoggedCatch<Unit, UpdateManager, UnauthorizedAccessException>(this, _ => Observable.Return(Unit.Default)))
+                .SelectMany(x => Utility.DeleteDirectory(x.FullName))
```
Comment (shiftkey) [inline src/Shimmer.Client/UpdateManager.cs]: One last thing - i'll let callers pass in an `IScheduler` here and drop the need to do the subsequent `ObserveOn` below

**Pull request merged**